### PR TITLE
Mini Project 00

### PR DESCRIPTION
**Pull request by cmot17**

### CODE REVIEW

File: mini-projects/00-jq-rs/.idea/.gitignore
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/.idea/.gitignore:1]: Extraneous files in PR :( (-1)

File: mini-projects/00-jq-rs/src/main.rs [hunk 1]
```diff
@@ -1,3 +1,277 @@
...
+    let args = Args::parse();
+
+    // Determine if color should be used
+    let use_color = if args.monochrome_output {
+        false
+    } else {
+        args.color_output
+    };
+
+    let input = BufReader::new(std::io::stdin());
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/main.rs:52]: Supposed to use file input, not stdin (-0.5)

File: mini-projects/00-jq-rs/src/main.rs [hunk 2]
```diff
@@ -1,3 +1,277 @@
...
+#[derive(Parser, Debug)]
+#[command(version, about, long_about = None)]
+struct Args {
+    filter_string: String,
+
+    #[arg(
+        short = 'C',
+        long,
+        default_value_t = true,
+        conflicts_with = "monochrome_output"
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/main.rs:20]: Great job leveraging the capabilities of clap here - I didn't actually know `conflicts_with` exists the first time I did this assignment

File: mini-projects/00-jq-rs/src/parse.rs [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+                    }
+                }
+                _ => Err(ParseError::UnexpectedToken(Token::LeftParenthesis)),
+            }
+        }
+        _ => Err(ParseError::UnexpectedEndOfInput),
+    }
+}
+
+fn parse_array_indexes<I>(iter: &mut I) -> Result<Vec<usize>, ParseError>
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/parse.rs:119]: Something about this function isn't quite working...

Concretely speaking, it's (-1) for the `del` function not working for slices (doesn't seem to parse the colon properly).

I'd probably try and refactor this logic so you have index parsing done correctly in one place for both the array index/slice filters and the del filter, since the former does handle slices correctly

File: mini-projects/00-jq-rs/src/tokenizer.rs
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/tokenizer.rs:1]: Love the tokenizer - I mentioned this previously but this is essentially a mini-version of the first step of writing a compiler

File: mini-projects/00-jq-rs/src/parse.rs [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+{
+    match iter.next() {
+        Some(Token::Number(n)) => Ok(n),
+        Some(token) => Err(ParseError::UnexpectedToken(token)),
+        None => Err(ParseError::UnexpectedEndOfInput),
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/parse.rs:186]: Your parsing unit test coverage is good, but there's a bit of room for improvement - for instance, no failing test case for deleting an array slice.

Also, not seeing parsing unit tests for iterator, add, and len.

File: mini-projects/00-jq-rs/src/main.rs [hunk 5]
```diff
@@ -1,3 +1,277 @@
...
+            use_color,
+            args.compact_output,
+        );
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/main.rs:78]: Again, test coverage is pretty good, but could do with some more failing test cases for things like `object identifier` and `slices`. This is lumped with parsing unit tests in the rubric, so I'll dock (-4) for the two combined.

File: mini-projects/00-jq-rs/src/print.rs
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/print.rs:1]: Printer code looks fine - a little messy, but it is a messy problem, so not too much sweat here. 

Minor efficiency tidbit from the firmware guy on static vs dynamic dispatch - In this case, because your final application is only actually using a single implementation of `Write` - `io::stdin` - dynamic dispatch is not actually saving any binary size compared to static dispatch, but you still incur the performance cost of the vtable lookup. Does this realistically matter? No, not really, but it is fun :)

**Review by amit-kumarh — COMMENTED** [review]
# Rubric For Mini Project 0: JQ

## Submission

### Canvas

1: The link submitted on Canvas is for the pull request containing the
assignment submission.

0: The link submitted on Canvas is not for the pull request containing the
assignment submission.

### Git Branch

1: The branch includes only changes for this assignment.

0: The branch includes commits or changes outside this assignment.Extrr

### Git Commit Message

1: All commit messages for the branch are reasonably clear.

0: One or more commit messages for the branch are reasonably clear.

### GitHub Pull Request

1: The pull request is made to the `main` branch of the student's fork of the
course repository.

0: The pull request is made to a branch other than `main` or to a repo that is
not the student's fork of the course repository.

### Points

**Submission Score: 3/4**

## Features
### Command Line Interface
For each of the required CLI arguments `filter-string`, `path-to-json`, `color-output`, `monochrome-output`, `sort-keys`, `indent`, and `compact-output`

0.5: The argument can be correctly parsed and consumed by the rest of the package.

0: The argument is not correctly parsed and propagated to the rest of the package.

Additionally, the command line parsing correctly raises errors to the user when conflicting arguments are inputted.

1.5: Errors are properly determined and bubbled up to the user.

0: Errors are not propagated back to the user.

### Supported Filters
For each of the required filters `identity`, `object-identifier index`, `array index`, `array slice`, `pipe`, and `array iterator`

2: The filter is correctly implemented and properly parses all possible JSON inputs.

1: Minor errors exist in the implement of the filter.

0: Major errors exist in the implementation, or the filter was omitted from the package.

Additionally, comprehensive unit tests are required for eachMinor  feature:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the filter, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the filter.

0: Tests are missing, or lack significant assessment of the filter's performance.

### Built-In Functions
For each of the required built-in functions `add`, `length`, `del`

2: The built-in function is correctly implemented and properly processes all possible JSON inputs.

1: Minor errors exist in the implementation of the operator.

0: Major errors exist in the implementation, or the function was omitted from the package.

Additionally, comprehensive unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the function, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the function.

0: Tests are missing, or lack significant assessment of the function's performance.

### Formatted Printing
For each of the required formats, `colored printing`, `monochrome printing`, `sorted keys`, `indents`, and `compact output`

2: The formatter is correctly implemented and properly displays any JSON object..

1: Minor errors exist in the implementation of the formatter.

0: Major errors exist in the implementation, or the formatter was omitted from the package.

Additionally, unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to ensure that the formatter can correctly handle all expected inputs.

1: Unit tests are present, but lack significant breadth of cases to fully exercise the formatter.

0: Tests are missing, or lack significant assessment of the formatter.

### Points
**Features Score: 55.5/61**

## Code Quality

### Design

#### File/Module organization

2: The program is appropriately split into reasonable files/modules

1: The program is somewhat organized, but there's room for improvement (splits aren't idiomatic, or some modules are still too large, etc.)

0: The program is not split at all

#### Scalability

5: The code is architected with scalability in mind. If more features needed to be added to the filtering/printing/CLI, they could easily be integrated.

3: The code was architected with some scalaibility in mind. However, significant refactoring would be needed in order to expand the feature set of this tool. (Possible symptoms include not modularizing enough/not effectively using traits + generics).

0: The code is not scalable. The entire package would need to be refactored in order to expand its features.

The filter code is pretty good, but adding to the parser would be a little convoluted - seems like it'd require editing multiple enums/matches, and having to figure out where to integrate into the existing parsing code is not straightforward (-1)

#### General Cleanliness

4: The code is effectively documented, especially for sources of user-facing errors. Additionally, function signatures are clear and the code is overall very readable.

2: Attempts at verbose documentation were made, but certain sources of error are unclear OR poor function signatures are used and readers would have slight difficulties while scanning the code.

0: Documentation is severely lacking in several palces and the code is not written in a readable way.

Sparse comments are there, but I'd like to see more comprehensive use of RustDoc and adding docstrings for functions/types (-2). Some functions are also quite long and nested, but nothing too egregious.

### Lint/Format

#### RustFmt

2: RustFmt produces no warnings when run on the changed files

1: RustFmt produces a few minor warnings when run on the changed files

0: RustFmt produces many warnings when run on the changed files

#### Clippy

2: Clippy produces no warnings when run on the changed files

1: Clippy produces a few minor warnings when run on the changed files

0: Clippy produces many warnings when run on the changed files

### Points
**Code Quality Score: 12/15**

## Summary
Add up the point totals here

**Total Score: 70.5/80**

Nice job - see line level comments for some more details
